ark/jsonschema/composition.ts

```diff
@@ -0,0 +1,60 @@
+import { printable } from "@ark/util"
+import { type, type JsonSchema, type Type } from "arktype"
+import { parseJsonSchema } from "./json.ts"
+
+const parseAllOfJsonSchema = (jsonSchemas: readonly JsonSchema[]): Type =>
+	jsonSchemas
+		.map(jsonSchema => parseJsonSchema(jsonSchema))
+		.reduce((acc, validator) => acc.and(validator))
+
+const parseAnyOfJsonSchema = (jsonSchemas: readonly JsonSchema[]): Type =>
+	jsonSchemas
+		.map(jsonSchema => parseJsonSchema(jsonSchema))
+		.reduce((acc, validator) => acc.or(validator))
+
+const parseNotJsonSchema = (jsonSchema: JsonSchema) => {
+	const inner = parseJsonSchema(jsonSchema)
+	return type.unknown.narrow((data, ctx) =>
+		inner.allows(data) ?
+			ctx.reject({
+				expected: `a value that's not ${inner.description}`,
+				actual: printable(data)
+			})
+		:	true
+	) as Type
+}
+
+const parseOneOfJsonSchema = (jsonSchemas: readonly JsonSchema[]) => {
+	const oneOfValidators = jsonSchemas.map(nestedSchema =>
+		parseJsonSchema(nestedSchema)
+	)
+	const oneOfValidatorsDescriptions = oneOfValidators.map(
+		validator => `○ ${validator.description}`
+	)
+	return type.unknown.narrow((data, ctx) => {
+		let matchedValidator: Type | undefined = undefined
+
+		for (const validator of oneOfValidators) {
+			if (validator.allows(data)) {
+				if (matchedValidator === undefined) {
+					matchedValidator = validator
+					continue
+				}
+				return ctx.reject({
+					expected: `exactly one of:\n${oneOfValidatorsDescriptions.join("\n")}`,
+					actual: printable(data)
```

Comment:
ideally because this is such a potentially confusing constraint we might list all satisfied branches, but I don't care that much as I don't think this should be used anyways. thanks for implementing it though- I know it's in the spec🫡

Reply:
In theory I did this in b31364f, but it appears the error message just states `must be valid according to jsonSchemaOneOfValidator` rather than using my custom error info... I asked about this on Discord [here](https://discord.com/channels/957797212103016458/1196196794611286087/1355631004533919805)